sound/soc/sof/sof-audio.c

```diff
@@ -166,15 +166,339 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
+		list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+			if (w->comp_id == pipe_widget->comp_id) {
+				pipeline_found = true;
+				break;
+			}
+		}
+		if (!pipeline_found) {
+			/*
+			 * Pipeline widgets can only be set up after the scheduling
+			 * widget has been set up. Create a temp widget to save it to the spcm
+			 * pipeline list as the original widget is already part of the core's
+			 * widget list.
+			 */
+			temp_pipe_widget = kmemdup(pipe_widget, sizeof(*pipe_widget), GFP_KERNEL);
+			if (!temp_pipe_widget) {
+				ret = -ENOMEM;
+				goto widget_free;
+			}
+
+			list_add(&temp_pipe_widget->list, &spcm->stream[dir].pipeline_list);
+		}
+	}
+
+	/*
+	 * error in setting pipeline connections will result in route status being reset for
+	 * routes that were successfully set up when the widgets are freed.
+	 */
+	ret = sof_setup_pipeline_connections(sdev, list, dir);
+	if (ret < 0)
+		goto widget_free;
+
+	/* setup pipeline widgets and complete pipelines */
+	list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = snd_sof_pipeline_find(sdev, w->pipeline_id);
+
+		ret = sof_widget_setup(sdev, pipe_widget);
+		if (ret < 0)
+			goto pipe_free;
+
+		pipe_widget_count++;
+
+		if (pipe_widget->complete)
+			continue;
+
+		pipe_widget->complete = snd_sof_complete_pipeline(sdev->dev, pipe_widget);
+		if (pipe_widget->complete < 0) {
+			ret = pipe_widget->complete;
+			goto pipe_free;
+		}
+	}
+
+	return 0;
+
+pipe_free:
+	/* free the pipe widgets that were setup successfully set up */
+	list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+		struct snd_sof_widget *pipe_widget = snd_sof_pipeline_find(sdev, w->pipeline_id);
+
+		if (!pipe_widget_count--)
+			break;
+
+		if (pipe_widget)
+			sof_widget_free(sdev, pipe_widget);
+	}
+
+widget_free:
+	/* free all widgets that have been setup successfully */
+	for_each_dapm_widgets(list, j, widget) {
+		/* i is the count of the widgets that were setup successfully */
+		if (j == i)
+			break;
+
+		if (is_sof_widget(widget))
+			sof_widget_free(sdev, widget->dobj.private);
+	}
+
+	/*
+	 * free all the temp_widgets in the pipeline list including the ones that may not
+	 * have been set up yet
+	 */
+	list_for_each_entry_safe(w, _w, &spcm->stream[dir].pipeline_list, list) {
+		list_del(&w->list);
+		kfree(w);
+	}
+
+	return ret;
+}
+
+int sof_widget_list_free(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct sof_ipc_fw_version *v = &sdev->fw_ready.version;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w, *_w;
+	int i;
+
+	/* nothing to free */
+	if (!list)
+		return 0;
+
+	/* Don't free widgets if firmware ABI is older than 3.19 but send the DAI_CONFIG IPC */
+	if (v->abi_version < SOF_ABI_VER(3, 19, 0)) {
+		ret = sof_widget_list_dai_config_setup(sdev, list);
```

Comment:
why are we calling setup in this widget list free routine?

Reply:
so reason I'm asking, my multi_ssp patches don't apply cleanly anymore... where shoull I put the dai_config/pcm match now?

Reply:
@juimonen this is .mainly to send the updated link ID for HDA and SDW DAIs. For your case, you won't need to worry because this is only for older ABI